src/lib/classes/engine/EngineStorage.ts

```diff
@@ -0,0 +1,165 @@
+import type IEngineStorageObject from "$lib/interfaces/IEngineStorageObject";
+import { Engine } from "./Engine";
+import type { EngineType } from "./EngineType";
+/**
+ * Engine storage definition.
+ * Stores an array of Engine instances, and can create, delete or search the array
+ * */
+class EngineStorage {
+	/**
+	 * Array of engines defined
+	 * */
+	engineArray: Array<Engine> = [];
+
+	/**
+	 * variable used to give unique id to engines
+	 * */
+	#engineId: number = 0;
+
+	get engineId(): number {
+		return (this.engineId = ++this.#engineId); //auto inc
+	}
+	set engineId(id: number) {
+		if (id >= 0) this.#engineId = id;
+		else throw new Error("Invalid engineId");
+	}
+
+	/**
+	 * Array of engines defined
+	 * */
+	#defaultEngine: Engine | undefined;
+	get defaultEngine(): Engine | undefined {
+		return this.#defaultEngine;
+	}
+	set defaultEngine(engine: Engine | undefined) {
+		let index = -1;
+
+		index = this.engineArray.findIndex((engine: Engine) => {
+			return engine.id === engine.id;
+		});
+		if (engine === undefined || index > -1) this.#defaultEngine = engine;
+		else throw new Error("Failed to set default engine");
+	}
+
+	constructor() {
+		this.engineArray = new Array<Engine>();
+		this.engineId = 0;
+		this.defaultEngine = undefined;
+	}
+
+	/**
+	 * Create an Engine and pushes it to engineArray
+	 */
+	createEngine(
+		name: string,
+		address: string,
+		portRangeStart: number,
+		portRangeEnd: number,
+		type: EngineType,
+		useBundle: boolean,
+	) {
+		const newEngine = new Engine(
+			name,
+			address,
+			portRangeStart,
+			portRangeEnd,
+			type,
+			this.engineId,
+			useBundle,
+		);
+		this.engineArray.push(newEngine);
+	}
+
+	/**
+	 * Deletes the engine with the given id
+	 */
+	deleteEngine(id: number) {
+		const engineIndex: number = this.engineArray.findIndex(
+			(engine: Engine) => {
+				return engine.id === id;
+			},
+		);
+		//check for results
+		if (engineIndex > -1) {
+			//Check if the default engine is being removed
+			if (this.defaultEngine === this.engineArray[engineIndex])
+				this.defaultEngine = undefined;
+
+			this.engineArray.splice(engineIndex, 1);
+		} else {
+			throw new Error("Engine Id does not exist");
+		}
+	}
```

Comment:
Why is a function like this so complicated? If all engines have numeric IDs, put the engine in the array at the index that corresponds to their ID. Then when you want to delete the engine, just set that index to `undefined`.